### PR TITLE
Improve serverside performance

### PR DESCRIPTION
**Pull request by Python1320**
Do not iterate over a huge array, but make the runtime checking work through time keeping.
Pending proper testing and validation. All I'm worried of is that this makes output firing stricter.

EDIT by TomyLobo:
old PR: https://github.com/wiremod/wire/pull/1021

### CODE REVIEW

**Comment by thegrb93** [comment]
It works but only allows two triggers per frame. Idk if that's good or bad (good in my opinion, but it used to be 8).

I used this starfall code to test it using an increment gate, and it only reaches 2.
```
if CLIENT then return end

wire.adjustOutputs({"Reset","Add"},{"normal","normal"})

wire.ports.Reset = 1
wire.ports.Reset = 0

for I = 1, 100 do
    wire.ports.Add = 1
    wire.ports.Add = 0
end
```

**Comment by Python1320** [comment]
Could you try adjusting the constant to see if you can reach at least around 4 per frame, I'll update the pull request with the new constant then. (I think the original limit is 4 at maximum if you look at the original hook code. Limit 8 happened only on first frame of output creation, which I hope does not matter.)

This is the most expensive hook on our server (at certain times) according to our profiler which is why I've tried to get it optimized.

**Comment by thegrb93** [comment]
I just remembered that each increment is actually two triggers, so 2 increments is 4 triggers which is expected it seems. I'll test it with a think hook and see if it's working each tick.

**Comment by thegrb93** [comment]
Thanks for pointing out that hook in starfall. I'll be fixing that since I think it's what is breaking the spawning on my server. It will be on my branch so follow if you want, I have added a lot of stuff since the original devs don't develop for it anymore (and they banned me).

**Comment by Nebual** [comment]
Very neat idea, I approve.

Though that 1/0 is gross and silly.

**Comment by thegrb93** [comment]
1/0?

**Comment by Python1320** [comment]
Now using math.huge and lowered the limit back down to 5 instead of 8 or 4

File: lua/wire/server/wirelib.lua [hunk 1]
```diff
@@ -19,6 +19,8 @@ local tostring = tostring
 local Vector = Vector
 local Color = Color
 local Material = Material
+local FrameTime = FrameTime
+local CurTime= CurTime
```
Comment (TomyLobo) [inline lua/wire/server/wirelib.lua]: missing space

**Comment by TomyLobo** [comment]
This needs more comments. I definitely can't see what's going on by looking at the code for a minute.

**Comment by Python1320** [comment]
This appears to be completely wrong right now according to my simulations anyway. (The old behavior is not being replicated)

I'm trying to remake this by tweaking some constants, I've gotten some promising results:


Simulator code:
```lua
local frametime = FrameTime()
local cur_time = CurTime()
local max_overtime = math.huge


local TriggerLimit = cur_time

local fire = function()
	local lastfire = TriggerLimit
	if (lastfire >= max_overtime) then Msg"_ " return end
	TriggerLimit = (lastfire < cur_time and cur_time or lastfire) + frametime*(1/4.5)
	Msg"X "
end

local x=0
hook.Add("Think", "test", function()
	x=x+1
	if x>20 then 
		hook.Remove("Think","test")
		return 
	end
		
	frametime = FrameTime()
	cur_time = CurTime()
	max_overtime = cur_time + frametime
	if x>9 and x<12 then Msg"not firing, " return elseif x==12 then MsgN"" end
	Msg"Firing... "
	for i=1,35 do fire() end
	MsgN" Fired"
end)
```

Current results:
```lua
[Luadev Server] <0:0:13073749|Python1320><New 9> running on server
Firing... X X X X X _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _  Fired
Firing... X X X X _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _  Fired
Firing... X X X X X _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _  Fired
Firing... X X X X _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _  Fired
Firing... X X X X X _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _  Fired
Firing... X X X X _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _  Fired
Firing... X X X X X _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _  Fired
Firing... X X X X _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _  Fired
Firing... X X X X X _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _  Fired
not firing, not firing, 
Firing... X X X X X _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _  Fired
Firing... X X X X _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _  Fired
Firing... X X X X X _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _  Fired
Firing... X X X X _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _  Fired
Firing... X X X X X _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _  Fired
Firing... X X X X _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _  Fired
Firing... X X X X X _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _  Fired
Firing... X X X X _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _  Fired
Firing... X X X X X _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _  Fired
```
I can even add bursting if needed after inactivity.

**Comment by Nebual** [comment]
I've been testing this ingame, by directly running TriggerOutput on an Identity Gate that's wired to another gate, and looking at it's input.
`lua_run for i=0, 20, 1 do WireLib.TriggerOutput(Entity(1234), 'Out', i) end`

Before this patch I consistently have it trigger 4 times per run (ending up with a display value of 3 on all subsequent runs, since 3->0, 0->1, 1->2, 2->3 is 4 changes). After this patch it seems to consistently trigger 5 times per run. I am on a 66 tick server, if that matters. Why are you fiddling with floats, which seems to be causing that 4,5,4,5,4,5 pattern you pasted? Was the old behaviour not consistently 4 triggers?

**Comment by Nebual** [comment]
@Python1320 Any news on this one? It seemed like a good idea.

**Comment by AbigailBuccaneer** [comment]
Seconded, I'd like to see this go ahead (and be consistent).

**Comment by Python1320** [comment]
@AbigailBuccaneer @Nebual it should be good to go. Used the latest version on meta for quite a while and no complaints. 

The 4,5,4 pattern is to match or exceed previously allowed limit.  I don't remember specifics anymore but it shouldn't affect anything unless you're hitting the limit on purpose that is.

**Comment by AbigailBuccaneer** [comment]
Bringing this back up again. I like the performance gains, and not doing useless work, but I don't like the nondeterminism and potential tickrate-specific issues.

What if each output stored the time of last trigger as well as a how many triggers it has left in the tick? That way we still have a deterministic trigger limit, and we don't need a think hook to iterate over every output. The only downside is that `WireLib.TriggerOutput` has to do marginally more work and we have to store an extra field.

**Pull request closed without merging**